File: planner-prototype/server/src/main/java/kpn/planner/repository/GraphRepositoryImpl.java
package kpn.planner.repository;

import java.util.ArrayList;
import java.util.List;

import org.ektorp.CouchDbConnector;
import org.ektorp.ViewQuery;
import org.ektorp.ViewResult;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import kpn.planner.domain.NetworkType;
import kpn.planner.domain.document.JsonNodeMapper;
import kpn.planner.domain.document.RouteDocument;
import kpn.planner.util.Databases;

@Repository
public class GraphRepositoryImpl implements GraphRepository {

	private final CouchDbConnector database;

	@Autowired
	public GraphRepositoryImpl(Databases databases) {
		this.database = databases.getMain();
	}

	public RouteDocument getRouteDocumentById(String routeId) {
		return database.get(RouteDocument.class, "route:" + routeId);
	}

	public List<JsonNodeMapper> getNodesFromFiles(NetworkType networkType) {

		String startkey = String.format("[\"%s\"]", networkType.getTagKey());
		String endkey = String.format("[\"%s\",{}]", networkType.getTagKey());

		ViewQuery query = new ViewQuery();
		query.designDocId("_design/PlannerDesign");
		query.viewName("GraphEdgesView");
		query.cacheOk(true);
		query.queryParam("reduce", "false");
		query.queryParam("stale", "ok");
		query.queryParam("startkey", startkey);
		query.queryParam("endkey", endkey);

		ViewResult result = database.queryView(query);

		return parseViewResult(result);
	}

	private List<JsonNodeMapper> parseViewResult(ViewResult result) {
		List<JsonNodeMapper> nodeList = new ArrayList<>();
		ObjectMapper objectMapper = new ObjectMapper();

		result.getRows().forEach(row -> {
			JsonNodeMapper nodeMapper = new JsonNodeMapper();
			nodeMapper.setKey(objectMapper.convertValue(row.getKeyAsNode(), new TypeReference<ArrayList<String>>() {
			}));
			nodeMapper.setValues(objectMapper.convertValue(row.getValueAsNode(), new TypeReference<ArrayList<Long>>() {
			}));
			nodeList.add(nodeMapper);
		});

		return nodeList;
	}
}

File: server/src/main/scala/kpn/core/repository/GraphRepositoryImpl.scala
package kpn.core.repository

import kpn.core.db.couch.Couch
import kpn.core.db.couch.Database
import kpn.core.db.views.GraphEdgesView
import kpn.core.planner.graph.GraphEdge
import kpn.shared.NetworkType
import spray.http.Uri
import spray.http.Uri.Query

class GraphRepositoryImpl(database: Database) extends GraphRepository {

  override def edges(networkType: NetworkType): Seq[GraphEdge] = {

    val uri = Uri(s"_design/PlannerDesign/_view/GraphEdgesView?reduce=false&stale=ok&startkey=[${networkType.name}]&endkey=[${networkType.name},{}]")

    val parameters = Map(
      "startkey" -> s"""["${networkType.name}"]""",
      "endkey" -> s"""["${networkType.name}",{}]""",
      "reduce" -> "false",
      "stale" -> "ok"
    )

    val request = uri.withQuery(Query(parameters))
    database.getRows(request.toString(), Couch.batchTimeout).map(GraphEdgesView.convert)
  }
}
